File: library/src/main/java/sk/vander/library/BaseUiModule.java
package sk.vander.library;

import android.app.Activity;

import dagger.Module;
import dagger.Provides;
import sk.vander.library.annotation.ActivityScreenSwitcherServer;
import sk.vander.library.annotation.ApplicationScope;
import sk.vander.library.navigation.activity.ActivityScreenSwitcher;

/**
 * Created by arashid on 21/06/16.
 */
@Module
public class BaseUiModule {

  @Provides @ApplicationScope ActivityScreenSwitcher provideActivityScreenSwitcher() {
    return new ActivityScreenSwitcher();
  }

  @Provides @ApplicationScope @ActivityScreenSwitcherServer
  ActivityHierarchyServer provideActivityScreenSwitcherServer(final ActivityScreenSwitcher screenSwitcher) {
    return new ActivityHierarchyServer.Empty() {
      @Override
      public void onActivityStarted(Activity activity) {
        screenSwitcher.attach(activity);
      }

      @Override
      public void onActivityStopped(Activity activity) {
        screenSwitcher.detach();
      }
    };
  }
}

File: library/src/main/java/sk/vander/library/BaseUiModule.kt
package sk.vander.library

import android.app.Activity

import dagger.Module
import dagger.Provides
import sk.vander.library.annotation.ActivityScreenSwitcherServer
import sk.vander.library.annotation.ApplicationScope
import sk.vander.library.navigation.activity.ActivityScreenSwitcher

/**
 * Created by arashid on 21/06/16.
 */
@Module
class BaseUiModule {

  @Provides @ApplicationScope fun provideActivityScreenSwitcher(): ActivityScreenSwitcher {
    return ActivityScreenSwitcher()
  }

  @Provides
  @ApplicationScope
  @ActivityScreenSwitcherServer
  fun provideActivityScreenSwitcherServer(screenSwitcher: ActivityScreenSwitcher): ActivityHierarchyServer {
    return object : ActivityHierarchyServer.Empty() {
      override fun onActivityStarted(activity: Activity) {
        screenSwitcher.attach(activity)
      }

      override fun onActivityStopped(activity: Activity) {
        screenSwitcher.detach()
      }
    }
  }
}